store/iterator.go

```diff
@@ -52,33 +62,37 @@ func ascendBtree(bt *btree.BTree, start, end []byte) *btreeIter {
 		} else { // both != nil
 			bt.AscendRange(bkey{start}, bkey{end}, insert)
 		}
-		iter.onceRead.Do(func() { close(read) })
 	}()
 
 	return iter
 }
 
 func descendBtree(bt *btree.BTree, start, end []byte) *btreeIter {
 	read := make(chan btree.Item)
-	// ensure we never block when we call close()
-	stop := make(chan struct{}, 1)
+	stop := make(chan struct{})
+	released := make(chan struct{})
 	iter := &btreeIter{
 		read:      read,
 		stop:      stop,
+		released:  released,
 		ascending: false,
 	}
 
-	insert := func(item btree.Item) bool {
-		select {
-		case read <- item:
-			return true
-		case <-stop:
-			iter.onceRead.Do(func() { close(read) })
-			return false
+	go func() {
+		defer func() {
+			close(read)
+			close(released)
+		}()
+
+		insert := func(item btree.Item) bool {
```

Comment:
Can you pull the insert function above the goroutine to make the goroutine logic a bit clearer?

Reply:
Continued in https://github.com/iov-one/weave/pull/859#discussion_r300625626